### PR TITLE
osc expose: Expose routes from services

### PR DESCRIPTION
**Pull request by kargakis**
This commit adds the expose command from Kubernetes with a couple of additions. `osc expose` like `kubectl expose` exposes services, pods, or replication controllers as services and then it goes on exposing deployment configs as services and services as routes.

In order for `osc expose` to generate routes from services, routes had to implement the kubectl.Generator interface.

cc: @smarterclayton

### CODE REVIEW

**Comment by kargakis** [comment]
This is a prerequisite: https://github.com/openshift/origin/pull/1941

**Comment by kargakis** [comment]
The issue with `expose` requiring ports is now fixed. Everything works like a charm. @smarterclayton PTAL

**Comment by kargakis** [comment]
@smarterclayton addressed all of your comments. Not sure about the service validation though (cc: @pweil- )

File: pkg/cmd/util/clientcmd/factory.go [hunk 1]
```diff
@@ -106,6 +107,15 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
 		return describe.NewHumanReadablePrinter(noHeaders), nil
 	}
 
+	generators := map[string]kubectl.Generator{
+		"route/v1": routegen.RouteGenerator{},
```
Comment (kargakis) [inline pkg/cmd/util/clientcmd/factory.go]: @smarterclayton do you want `osc expose` to support exposing services from services or just routes?

Comment (smarterclayton) [inline pkg/cmd/util/clientcmd/factory.go]: I think routes.  I think we'd want to have a copy or create-external-service function for service from service.

----- Original Message -----

**Comment by kargakis** [comment]
The last PR upstream is going to be merged soon so I guess this is ready for a review. Behavior of the command can be found in the first post.
@smarterclayton @pweil-

**Comment by smarterclayton** [comment]
```
++ Building go targets for linux/amd64: cmd/openshift
# github.com/openshift/origin/pkg/cmd/cli
_output/local/go/src/github.com/openshift/origin/pkg/cmd/cli/cli.go:59: undefined: f
_output/local/go/src/github.com/openshift/origin/pkg/cmd/cli/cli.go:59: undefined: out
```

**Comment by kargakis** [comment]
I've refactored most of `cli.go` in the stop|label PR and I don't want to merge myself, that's why this doesn't run.

**Comment by kargakis** [comment]
[test]

**Comment by kargakis** [comment]
@smarterclayton tests pass. Please review or merge this if everything's ok.

**Comment by smarterclayton** [comment]
So I'm a bit confused as to why this can't expose replication controllers or deployment configs?  We don't want to completely change the behavior of expose - we want to supplement it.  "expose"ing a service should be creating a route, but "expose"ing everything else should create a service.  Maybe I misread the pull, but that doesn't look like what the code is doing?

**Comment by kargakis** [comment]
> So I'm a bit confused as to why this can't expose replication controllers or deployment configs? We don't want to completely change the behavior of expose - we want to supplement it. "expose"ing a service should be creating a route, but "expose"ing everything else should create a service. Maybe I misread the pull, but that doesn't look like what the code is doing?

I thought we just wanted to expose routes from services and nothing else. Will add the ability to expose those resources you mentioned as services too.

**Comment by smarterclayton** [comment]
Sorry, I should have made clear that I was only referring to us overriding the behavior of expose of a service as a route, but preserving others.

**Comment by kargakis** [comment]
this is ready I guess

**Comment by kargakis** [comment]
@smarterclayton comments addressed

**Comment by smarterclayton** [comment]
LGTM [merge]

**Pull request merged**